Initial implementation of http namespace resolver

miminar opened the pull request:
Added a base generic resolver allowing for decent configuration.

TODOs:
- resolve all the small TODOs in code
- get the API of testing servers right
  - which codes are sent in case of which errors?
  - use json message for complex error reporting?

Resolves #495

Commit:
Initial implementation of http namespace resolver

Added a base generic resolver allowing for decent configuration.

TODOs:
- write more tests
- provide some default http resolver(s) based on generic
- use it as a base for another http resolver using config file with
  high priority entries
- resolve all the small TODOs in code
- get the API of testing servers right
  - which codes are sent in case of which errors?
  - use json message for complex error reporting?

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
Hi Derek, is this the right place for such a pull request? Or shall I submit it at `docker/distribution:master` branch with your work included?

Commit:
NSResolveActionCallback receives name being resolved

Let the name being resolved help in decision of what to do with
namespace extension.

Call the callback also on namespace entry containing no arguments. In
such a case the `namespace` argument will be empty.

Set logging level of logrus framework to `DEBUG` if running in verbose
mode.

Fixed cache accessing.

Added tests for hitting cache and bad cases.

Signed-off-by: Michal Minar <miminar@redhat.com>

dmcgowan commented:
Yes, here is fine for review. I will ask you to make a new pull request against master when namespaces are merged in

Commit:
Use config struct instead of multiple arguments to NewHTTPResolver

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Added option to ignore discovery failures

Only errors occuring during processing of namespace extensions can be
ignored.

Signed-off-by: Michal Minar <miminar@redhat.com>

dmcgowan commented on namespace/http_resolver.go:
```diff
@@ -0,0 +1,423 @@
+func (hr *httpResolver) resolveEntries(es *Entries, visited map[string]struct{}, name string) error {
+	cached, exists := hr.cache[name]
+	if exists && !cached.created.Add(hr.config.ExpireAfter).Before(time.Now()) {
+		entries, err := hr.config.ResolverFactory(cached.entries).Resolve(name)
```
Chaining together resolvers should not be done by this implementation.  Rather the `NewMultiResolver` can be used to compose multiple resolvers together. This prevents each resolver from needing to know how to join entries and stay focused on a simple lookup.

dmcgowan replied:
Think I misunderstood originally, this is intended to swap out the simple resolver for the cached use case right?

miminar replied:
Yeah, kind of. It has a confusing name for that purpose though. Now I think it would be better to use the same resolver both for cached and currently fetched entries.

The thing is I'd like to use `simpleResolver`'s resolve algorithm on fetched/cached entries rather than duplicate it here. Neither `multiResolver` nor `extendResolver` provide a way to pass entries obtained from one resolver to another and run `Resolve()` on them. Maybe some sort of:

```
type ResolverFactory func(*Entries) Resolver
type recursiveResolver struct {
    baseResolver Resolver
    resolverFactories []ResolverFactory
}

func (cr *recursiveResolver) Resolve(name string) (*Entries, error) {
    entries, err := cr.baseResolver.Resolver(name)
    if err != nil {
        return nil, err
    }
    for _, f := cr.resolverFactories {
        entries, err := f(entries).Resolve(name)
        if err != nil {
            return nil, err
        }
    }
    return entries, err
}
```

would do the trick. So I could write:

```
resolver := NewRecursiveResolver(NewHTTPResolver(...), func(entries *Entries) Resolver { return NewSimpleResolver(entries) })
```

WDYT?

dmcgowan commented on namespace/http_resolver.go:
```diff
@@ -0,0 +1,423 @@
+}
+
+func (hr *httpResolver) resolveEntries(es *Entries, visited map[string]struct{}, name string) error {
+	cached, exists := hr.cache[name]
```
The cache component could be separated into its own type which just caches the results of another resolver.

miminar replied:
Good idea. Will do.

miminar replied:
On the second thought, this will prevent `httpResolver` from reusing cached entries during recursion. If we have:

```
multiResolver{
    resolvers: []Resolver {
        cacheResolver{ resolver: &httpResolver{...} },
        cacheResolver{ resolver: &httpResolver{...} },
        cacheResolver{ resolver: &httpResolver{...} },
        ...
    }
}
```

there are separate caches. So even though the namespace entry being resolved at the 3rd level of recursion is stored in the 1st level cache, it will be fetched again.

Now I've got an idea. What about shared cache?

```
func NewCacheResolver(baseResolver Resolver, cache *SharedCache) (Resolver, SharedCache) {
    if cache == nil {
        cache := newSharedCache()
    }
    ...
    return resolver, cache
}

...

r, cache := NewCacheResolver(NewHttpResolver(...), nil)
NewMultiResolver(r, NewCacheResolver(NewHttpResolver(...), cache), ...)
```

Hmm, this bring me to a solution without any shared cache:

```
multiResolver{
    resolvers: []Resolver {
        // the same cache resolver all over again
        cacheResolver, cacheResolver, cacheResolver, ...
    }
}
```

But I think the concept of shared cache would be useful. I'll give it a try. Please stop me if you have better idea.

dmcgowan replied:
I am not sure I quite understand the concept of shared cache. In the multi resolver case the resolvers are treated like a chain and each may have a different view of the namespace depending on where they live in the chain. If cache is injected into that process and share data, then the view from one cache could show up at a different level that was not intended to have that view. It does seem complicated but I think it is easy to understand as a chain with first resolvers having priority.

dmcgowan commented on namespace/http_resolver.go:
```diff
@@ -0,0 +1,423 @@
+		return err
+	}
+
+	// handle scope extensions
```
The recursion shouldn't be necessary here.  Since it is common functionality it should be done in either the MultiResolver or another implementation which is able to recurse.

miminar replied:
Yeah, I see the point now. Originally, I wanted to use the cache during the recursion and thought this is the only way. Now it seems that cache can be decoupled from `httpResolver` while still having its benefits.

dmcgowan commented on namespace/http_resolver.go:
```diff
@@ -0,0 +1,423 @@
+				}
+				scp, newEntries, err := parseHTMLMetaTag(z, name)
+				if err != nil {
+					return nil, err
```
There will need to be support for the meta tags peacefully coexisting with unrelated meta tags

miminar replied:
Ok, I'll ignore unrecognized.

dmcgowan commented:
Thank you for implementing to spec. Since this is the first attempt at implementing it, do you see any changes that should be necessary? I had a conversation with @stevvooe before and he suggested possibly having the content of the meta tag look similar to an entry line in the plain text file. WDYT?

Also I think you exposed a need to break up the resolver into CacheResolver, and HTTPResolver, and possibly a RecursiveResolver.  Although ideally the logic for the recursive resolver would work with the mutli resolver implementation, which is attempting to interpret the namespace action.

Commit:
Decoupled cache and http resolver

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Removed recursion abilities from http resolver

Multi resolver should be used with it instead.

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Don't bail out on custom meta tag

Signed-off-by: Michal Minar <miminar@redhat.com>

miminar commented:
@dmcgowan So the discovery endpoint would reply with something like:

```
<meta name="docker-scope-entry" content="example.com push https://registry.example.com/v2/ version=2.0 trim">
<meta name="docker-scope-entry" content="example.com pull https://registry.example.com/v1/ version=1.0">
<meta name="docker-scope-entry" content="example.com push https://registry.mirror.com/v2/ version=2.0">
<meta name="docker-scope-entry" content="example.com pull http://registry.mirror.com/v2/ version=2.0">
<meta name="docker-scope-entry" content="example.com index https://search.mirror.com/v1/ version=1.0">
```

? It would simplify client's parser a lot. But I'm not sure if tying content's value with local configuration format is a good idea. Local configuration may later have other entries (namespace-less targets, aliases, custom attributes, etc.). Making them same from the beginning may lead to an impression that they are interchangeable. But, for example, I wouldn't want discovery endpoint to define aliases.

So I believe there's a benefig in having tags like `docker-registry-push`, `docker-index`, etc., parsing them and ignoring all the others.

miminar commented:
Hopefully I've covered your points. But there are some shortcomings to current implementation:
- Any failed attempt to contact momentarily unreachable discovery endpoint results in a fail. IMHO there should be a way to tell recursive http resolver to ignore errors occurring on extension namespaces.
- No easy way to define recursive http resolver. In tests I do it this way:
  
  ```
  const recursionLimit = 10
  resolvers := make([]Resolver, recursionLimit)
  hr := NewHTTPResolver(&HTTPResolverConfig{Client: newMockHTTPClient()})
  for i := 0; i < recursionLimit; i++ {
      resolvers[i] = hr
  }
  r := NewMultiResolver(resolvers...)
  ```
  
  Does it justify an addition of new `recursiveResolver` in addition to `multiResolver`?

dmcgowan commented:
Thinking...thinking...an additional `recursiveResolver` may make sense for the reason you specified. Either that or `multiResolver` should be updated to begin recursion of a namespace element at the current resolver, instead of the next.

dmcgowan commented on namespace/cache_resolver.go:
```diff
@@ -0,0 +1,123 @@
+		if sc.expireAfter == 0 || !entry.created.Add(sc.expireAfter).Before(time.Now()) {
+			return entry.entries
+		}
+		sc.garbageCollectExpired()
```
Trying to understand the intent of the garbage collection.  It seems this function is called when an expired entry is hit, however what is the value other than keeping the cache trimmed? Store will already evict an element if room is needed. Calling this function here could cause inconsistent runtime when clean up is needed.

Have you considered an LRU approach as well, such from https://github.com/hashicorp/golang-lru?

miminar replied:
I haven't considered it until now. The LRU approach IMHO isn't appropriate for caching http responses. It would prevent the most frequently searched items from ever being removed. If a configuration change occured at discovery endpoint, Docker daemon would stay oblivious to that for quite a while - until restarted. That's what this expiring cache is trying to solve.

miminar replied:
There's no other value than keeping the cache trimmed. I believe there's no need for keeping received entries once they are expired - they won't ever be used again. I'm unsure what you mean by inconsistent runtime.

dmcgowan commented on namespace/cache_resolver.go:
```diff
@@ -0,0 +1,123 @@
+	}
+	entry := newCacheEntry(name, entries)
+	sc.cache[name] = entry
+	sc.expirationQueue.PushBack(&entry)
```
There is no protection from store being called on a name that already exists in the cache. This could lead to premature deletion of the element.

miminar replied:
Good catch!

dmcgowan commented:
I would advocate for the underlying cache implementation to be made generic, such as https://github.com/hashicorp/golang-lru. But right now the cache implementation is not important, I would create a simple CacheResolver and maybe a cache interface which can be implemented by either a simple `map[string]interface{}`, your garbage collection implementation, or an LRU implemenation. The point is that it does not need to be decided here, but the interface and use should be clearly defined.

stevvooe commented on namespace/http_resolver_mock_test.go:
```diff
@@ -0,0 +1,272 @@
+}
+
+// TODO: are these something that would docker-distribution endpoint return
+func writeHeaders(w http.ResponseWriter, contentType string) {
```
I would just remove this. It shouldn't matter.

miminar replied:
Ok. Now setting only `Content-Type`.

Commit:
CacheResolver: Check for existing entry when adding new one

Also call garbage collector during both *lookup* and *store*.

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
HTTPResolver: Removed headers from test responses

Signed-off-by: Michal Minar <miminar@redhat.com>

Commit:
Provide cache interface for another cache implementation

Default implementation is ExpiringEntriesCache which removes oldest
entries when cache is filled up. It is targeted for use with http
resolver as a baseResolver.

Signed-off-by: Michal Minar <miminar@redhat.com>
